### DISCUSSION TITLE
How should I use the cluster-api-ipam-provider-in-cluster?

### DISCUSSION THREAD

**Discussion by lintest01** [kubernetes-sigs/cluster-api-ipam-provider-in-cluster#80] [Q&A] [answered]
I apologize for my lack of knowledge, but I am trying to use your cluster-api-ipam-provider-in-cluster and have encountered some issues. 
I created a cluster using the instructions on this page about Docker:
https://cluster-api.sigs.k8s.io/user/quick-start.html

Then I ran the following commands and confirmed that caip-in-cluster-system is running:
git clone https://github.com/telekom/cluster-api-ipam-provider-in-cluster
cd cluster-api-ipam-provider-in-cluster
kubectl apply -k config/default

Based on the README.md you provided, I created an inclusterippool-sample. Here are the specific details I used:
apiVersion: ipam.cluster.x-k8s.io/v1alpha1
kind: InClusterIPPool
metadata:
  name: inclusterippool-sample
spec:
  subnet: 10.0.0.0/24
  gateway: 10.0.0.1

Next, I tried to create a pod to verify that its IP address is within the inclusterippool, but it is clear that the pod's IP is not within the inclusterippool. Here is an example of the pod:
apiVersion: v1
kind: Pod
metadata:
  name: test-pod
spec:
  containers:
      - name: test-container
    image: nginx

I have reviewed the information you provided and did not see any indication that it can only be used in specific environments, such as AWS or vSphere.
 So, how can I verify that IPAM is functioning correctly? If there is any misunderstanding in the above content, please correct me. Alternatively, could you provide an example of how to use it so I can verify it? Thank you for your help.

**Comment by schrej** [comment, marked as answer]
Hey there. I think there is in fact a misunderstanding here. This IPAM provider is not intended for managing IP addresses of Pods within a cluster.
Cluster API is about deploying entire clusters, and this IPAM provider can be used to assign IP addresses to Machines (which end up as Nodes in a deployed cluster) that are deployed by Cluster API. Have a look at the [CAPI Proposal](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md) about the IPAM contract for more details.

It requires support from the infrastructure provider in order to be used. At the moment support is implemented in [CAPV](https://github.com/kubernetes-sigs/cluster-api-provider-vsphere), and I have an open PR for [CAPM3](https://github.com/metal3-io/cluster-api-provider-metal3) as well. It is most useful for on-premise providers, but _could_ also be used in a cloud environment.

**Reply by lintest01** [reply]
Thank you for letting me know. First, I understand that the Cluster API IPAM Provider In Cluster object is related to CAPV and its function. Regarding the content written in the IP Address Management Integration that you provided, I will take a closer look. Thank you for your guidance.

**Comment by xsen84** [comment]
Hi there
How exactly do i configure CAPV to use the In-Cluster IPAM?
I want to know where do i reference the inclusterippool.
Thank you

**Reply by schrej** [reply]
Sorry, I missed this. Hope you figured it out in the meantime. In case anyone finds it in the future: simply set `addressesFromPools` on the network device in the VSphereMachineTemplate.

https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/blob/main/apis/v1beta1/types.go#L359
